src/cache.c

```diff
@@ -0,0 +1,231 @@
+    void *value;
+    uint32_t key;
+    list_type_t arc_type;
+    struct list_head list;
+    struct list_head ht_list;
+} arc_entry_t;
+
+typedef struct hashtable {
+    struct list_head *ht_list_head;
+} hashtable_t;
+
+typedef struct cache {
+    struct list_head *list_table[4];
+    uint32_t list_size[4];
+    struct hashtable *map;
+    uint32_t capacity;
+    uint32_t lru_capacity;
+    uint64_t get_time;
+    uint64_t hit_time;
+} cache_t;
+
+cache_t *cache_create()
+{
+    cache_t *cache = malloc(sizeof(cache_t));
+    for (int i = 0; i < 4; i++) {
+        cache->list_table[i] = malloc(sizeof(struct list_head));
+        INIT_LIST_HEAD(cache->list_table[i]);
+        cache->list_size[i] = 0;
+    }
+    cache->map = malloc(sizeof(hashtable_t));
+    cache->map->ht_list_head = malloc(SIZE * sizeof(struct list_head));
+
+    for (int i = 0; i < SIZE; i++) {
+        INIT_LIST_HEAD(&cache->map->ht_list_head[i]);
+    }
+
+    cache->capacity = SIZE;
+    cache->lru_capacity = SIZE / 2;
+    cache->get_time = 0;
+    cache->hit_time = 0;
+    return cache;
+}
+
+/* Rule of ARC
+ * 1. size of LRU_list + size of LFU_list <= c
+ * 2. size of LRU_list + size of LRU_ghost_list <= c
+ * 3. size of LFU_list + size of LFU_ghost_list <= 2c
+ * 4. size of LRU_list + size of LFU_list + size of LRU_ghost_list + size of
+ * LFU_ghost_list <= 2c
+ */
+void assert_cache(cache_t *cache)
+{
+    assert(cache->list_size[LRU_list] + cache->list_size[LFU_list] <=
+           cache->capacity);
+    assert(cache->list_size[LRU_list] + cache->list_size[LRU_ghost_list] <=
+           cache->capacity);
+    assert(cache->list_size[LFU_list] + cache->list_size[LFU_ghost_list] <=
+           2 * cache->capacity);
+    assert(cache->list_size[LRU_list] + cache->list_size[LRU_ghost_list] +
+               cache->list_size[LFU_list] + cache->list_size[LFU_ghost_list] <=
+           2 * cache->capacity);
+}
+
+void move_to_mru(cache_t *cache, arc_entry_t *entry, const list_type_t arc_type)
+{
+    cache->list_size[entry->arc_type]--;
+    cache->list_size[arc_type]++;
+    entry->arc_type = arc_type;
+    list_del_init(&entry->list);
+    list_add(&entry->list, cache->list_table[arc_type]);
+}
+
+void replace_list(cache_t *cache)
+{
+    if (cache->list_size[LRU_list] >= cache->lru_capacity)
+        move_to_mru(
+            cache,
+            list_last_entry(cache->list_table[LRU_list], arc_entry_t, list),
+            LRU_ghost_list);
+    else if (cache->list_size[LFU_list] >=
+             (cache->capacity - cache->lru_capacity))
+        move_to_mru(
+            cache,
+            list_last_entry(cache->list_table[LFU_list], arc_entry_t, list),
+            LFU_ghost_list);
+}
+
+void *cache_get(cache_t *cache, uint32_t key)
+{
+    if (cache->capacity <= 0 ||
+        list_empty(&cache->map->ht_list_head[HASH(key)]))
+        return NULL;
+
+    arc_entry_t *entry = NULL;
+    list_for_each_entry(entry, &cache->map->ht_list_head[HASH(key)], ht_list)
```

Comment:
Update `.clang-format` to identify for-each macros. See https://github.com/sysprog21/lab0-c/blob/master/.clang-format